### PR TITLE
Add support for dotted field syntax

### PR DESCRIPTION
**Pull request by Gabriel439**
Fixes https://github.com/dhall-lang/dhall-lang/issues/220

This builds on top of #896 to add support for Nix/TOML/Cue-like dotted field
syntax of the form:

```dhall
{ x.y = 1, x.z = True }
```

... which is equivalent to:

```dhall
{ x = { y = 1, z = True } }
```

This is a useful feature in its own right (to simplify authoring large,
nested records), but will also pave the way for similar dotted notation for
record updates proposed in #754.

### CODE REVIEW

**Comment by Gabriel439** [comment]
Matching change to the Haskell implementation: https://github.com/dhall-lang/dhall-haskell/pull/1651

File: standard/dhall.abnf [hunk 1]
```diff
@@ -886,13 +885,19 @@ empty-record-literal = "="
...
+
+dotted-fields = any-label-or-some *("." any-label-or-some)
 
-non-empty-record-type = ":" whsp1 expression *(whsp "," whsp record-type-entry)
 record-type-entry = any-label-or-some whsp ":" whsp1 expression
 
-non-empty-record-literal = "=" whsp expression *(whsp "," whsp record-literal-entry)
-record-literal-entry = any-label-or-some whsp "=" whsp expression
+non-empty-record-literal =
+    dotted-fields "=" whsp expression *(whsp "," whsp record-literal-entry)
```
Comment (sjakobi) [inline standard/dhall.abnf]:
```suggestion
    dotted-fields whsp "=" whsp expression *(whsp "," whsp record-literal-entry)
```

… or shorter:

```suggestion
    record-literal-entry *(whsp "," whsp record-literal-entry)
```

File: standard/dhall.abnf [hunk 2]
```diff
@@ -886,13 +885,19 @@ empty-record-literal = "="
 empty-record-type = ""
 
 non-empty-record-type-or-literal =
-    any-label-or-some whsp (non-empty-record-literal / non-empty-record-type)
+    (non-empty-record-literal / non-empty-record-type)
+
+non-empty-record-type =
+    any-label-or-some whsp ":" whsp1 expression *(whsp "," whsp record-type-entry)
```
Comment (sjakobi) [inline standard/dhall.abnf]: Shorter:

```suggestion
    record-type-entry *(whsp "," whsp record-type-entry)
```

**Comment by Nadrieril** [comment]
Could you add a parser test for the case where dotted fields and duplicate fields interact ? For now you only have a normalization test
